spec/support/fake_ansible_repo.rb

```diff
@@ -0,0 +1,357 @@
+          version: master
+            name: nginx_role
+      REQUIREMENTS
+
+      # Content format/borrowed from the ansible documentation
+      #
+      #   https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#role-dependencies
+      META_DATA = <<~ROLE_DEPS
+        ---
+        dependencies:
+          - role: common
+            vars:
+              some_parameter: 3
+          - role: apache
+            vars:
+              apache_port: 80
+      ROLE_DEPS
+
+      VAR_DATA = <<~VARS
+        ---
+        var_1: foo
+        var_2: bar
+        var_3: baz
+      VARS
+
+      attr_reader :file_struct, :repo_path, :repo, :index
+
+      def self.generate repo_path, file_struct
+        new(repo_path, file_struct).generate
+      end
+
+      def initialize(repo_path, file_struct)
+        @repo_path   = Pathname.new(repo_path)
+        @name        = @repo_path.basename
+        @file_struct = parse_file_struct(file_struct)
+      end
+
+      def generate
+        FileUtils.mkdir_p(repo_path)
+        build_files_at(repo_path, file_struct)
+
+        git_init
+        git_add_all
+        git_commit_initial
+      end
+
+      # Create a new branch (don't checkout)
+      #
+      #   $ git branch other_branch
+      #
+      def git_branch_create(new_branch_name)
+        repo.create_branch(new_branch_name)
+      end
+
+      private
+
+      # Parses a multi-line string as a file structure definition and return an
+      # array of either single files or nested arrays/hashes.  The data
+      # structures are as follows:
+      #
+      #   - Array:  Collection of directory contents (either Strings or Hashes)
+      #   - String: A single filename
+      #   - Hash:   A collection (usually one) directory (key) with entries (value)
+      #
+      # The format can be delivered in two forms.  Either pure yaml:
+      #
+      #     - path/:
+      #       - to/:
+      #         - dir:
+      #           - foo
+      #           - bar
+      #           - baz
+      #     - other/:
+      #       - dir/:
+      #
+      # Or an plain text version that without some of the extra yaml chars added:
+      #
+      #     path/
+      #       to/
+      #         dir/
+      #           foo
+      #           bar
+      #           baz
+      #     other/
+      #       dir/
+      #
+      #
+      # Which is a cheeky trick with some regular expressions to allow for a
+      # more human readable format.  Both still go through yaml, and the extra
+      # formatting will be added only when missing.
+      #
+      # To use the second form, however, directories require a trailing slash,
+      # where they are optional in the yaml form (the `:` takes care of that).
+      #
+      def parse_file_struct(file_struct)
+        yaml_data = file_struct.gsub(/^(\s*)([a-zA-Z0-9\.])/, '\1- \2')
+                               .gsub(/\/$/, '/:')
+        YAML.load(yaml_data)
+      end
```

Comment:
I don't really like this as a testing interface.  We control both side of the equation, so if the final result is a Hash, I don't see why the callers can't just use a Hash.

```ruby
file_struct = {
  "path/" => {
    "to/" => {
      "dir/" => [
        "foo",
        "bar",
        "baz"
      ]
    }
  },
  "other/" => {
    "dir/": {}
  }
}
```

that isn't too far removed and it completely elimiates the YAML roundtripping.  Having giant hashes is a not-uncommon thing for most other specs in the test suite.

---

Additionally, do you need the trailing slash for directories?

Reply:
So this is where my "over engineered" comment from this morning came into play.  I agree and disagree with what you are saying to an extent, but will let you and Nick have the final say on what stays and what goes.  I will leave my 2 cents below this, but you are free to tell me to go shove it and change it if you don't want the parsing.

Oh, and regarding this:

> Additionally, do you need the trailing slash for directories?

Read my frick'n comment dude!  I said in the comment for the method:

```ruby
# To use the second form, however, directories require a trailing slash,
# where they are optional in the yaml form (the `:` takes care of that).
```

* * *

So the main reason motivation for even taking the time to do this was that when some else besides me ends up looking at these tests, I find it is much easier to decipher the plain text representation over the hash structure, or even the YAML form.

For example, take this comment for example (from Nick C.'s PR):

https://github.com/ManageIQ/manageiq/pull/19079#issuecomment-517392226

I feel like the structure provided in raw plain text:

```
ansible.cfg
playbooks/
  playbook1.yml
  playbook2.yml
common/
  roles/
    role1/
      tasks/
        main.yml
    role2/
      tasks/
        main.yml
```

Is much easier reason about than this:


```YAML
- ansible.cfg
- playbooks:
  - playbook1.yml
  - playbook2.yml
- common:
  - roles:
    - role1:
    - tasks:
        - main.yml
    - role2:
      - tasks:
        - main.yml
```

```ruby
[
  "ansible.cfg",
  { "playbooks" => [
    "playbook1.yml",
    "playbook2.yml"
  ]},
  { "common" => [
    { "roles" => [
      { "role1" => [
        "tasks" => [
          "main.yml"
        ]},
      ]},
    ]},
    { "roles" => [
      { "role2" => [
        "tasks" => [
          "main.yml"
        ]},
      ]},
    ]},
  }
]
```

Let alone type out, and I probably spent 5x-10x longer initially converting that from the other forms, and that doesn't count me having to toss that into my editor to fix the syntax errors I had...

But I do understand this argument:

> I just don't like the spec + String -> gsub | YAML.parse -> Hash when we can just do spec + Hash.

And admittedly this is just a "cool thing I wrote with a few lines of code, though I would say that the performance impact is pretty limited, and the main argument of the above is the supportability argument and that it is easily error prone as well.

Personally I think at a glance, the first is easier to understand, but just a preference.  Switching to using a raw hash isn't that hard to do.